common/src/View/SetBrushFaceAttributesTool.cpp

```diff
@@ -129,5 +138,103 @@ namespace TrenchBroom {
+
+            document->startTransaction("Drag Apply Face Attributes");
+
+            return true;
+        }
+
+        bool SetBrushFaceAttributesTool::doMouseDrag(const InputState& inputState) {            
+            const auto& hit = firstHit(inputState, Model::Brush::BrushHit);
+            if (!hit.isMatch()) {
+                // Dragging over void
+                return true;
+            }
+
+            Model::BrushFace* face = Model::hitToFace(hit);
+            if (face == m_dragTargetFace) {
+                // Dragging on the same face as last frame
+                return true;
+            }
+
+            if (m_dragSourceFace == nullptr && m_dragTargetFace == nullptr) {
+                // Start drag
+                m_dragSourceFace = m_dragInitialSelectedFace;
+                m_dragTargetFace = face;                
+            } else {
+                // Continuing drag onto new face
+                m_dragSourceFace = m_dragTargetFace;
+                m_dragTargetFace = face;
+            }
+
+            transferFaceAttributes(inputState, m_dragSourceFace, m_dragTargetFace);
+
+            return true;
+        }
+
+        void SetBrushFaceAttributesTool::doEndMouseDrag(const InputState&) {
+            auto document = kdl::mem_lock(m_document);
+            document->commitTransaction();
+
+            resetDragState();
+        }
+
+        void SetBrushFaceAttributesTool::doCancelMouseDrag() {
+            auto document = kdl::mem_lock(m_document);
+            document->cancelTransaction();
+
+            resetDragState();
+        }
+
+        void SetBrushFaceAttributesTool::resetDragState() {
+            m_dragInitialSelectedFace = nullptr;
+            m_dragTargetFace = nullptr;
+            m_dragSourceFace = nullptr;
+        }
+
+        void SetBrushFaceAttributesTool::transferFaceAttributes(const InputState& inputState, Model::BrushFace* sourceFace, Model::BrushFace* targetFace) {
+            ensure(sourceFace != nullptr, "no source face");
+            ensure(targetFace != nullptr, "no target face");
+            ensure(m_dragInitialSelectedFace != nullptr, "no initially selected face");
+
+            auto document = kdl::mem_lock(m_document);
+
+            const Model::WrapStyle style = wrapStyle(inputState);
+
+            const Transaction transaction(document);
+            document->deselectAll();
+            document->select(targetFace);
+
+            auto snapshot = sourceFace->takeTexCoordSystemSnapshot();
+            document->setFaceAttributes(sourceFace->attribs());
+            if (snapshot != nullptr) {
+                document->copyTexCoordSystemFromFace(*snapshot, sourceFace->attribs().takeSnapshot(), sourceFace->boundary(), style);
+            }
+
+            document->deselectAll();
+            document->select(m_dragInitialSelectedFace);
+        }
+
+        /**
+         * FIXME: Copied from SelectionTool, factor out
```

Comment:
What about this?

Reply:
I don't know, maybe it doesn't make sense to factor these out, but it does need to be documented. I can make another issue